src/python/pants/core/util_rules/environments.py

```diff
@@ -510,26 +510,32 @@ class EnvironmentSensitiveOptionFieldMixin:
 
 
 @memoized
-def add_option_fields_for(subsystem: type[Subsystem]) -> Iterable[UnionRule]:
-    """Register environment fields for the `environment_sensitive` options of `subsystem`
+def add_option_fields_for(env_aware: type[Subsystem.EnvironmentAware]) -> Iterable[UnionRule]:
+    """Register environment fields for the options declared in `env_aware`
 
-    This should be called in the `rules()` method in the file where `subsystem` is defined. It will
-    register the relevant fields under the `local_environment` and `docker_environment` targets.
+    This is called by `env_aware.subsystem.rules()`, which is called whenever a rule depends on
+    `env_aware`. It will register the relevant fields under the `local_environment` and
+    `docker_environment` targets. Note that it must be `memoized`, such that repeated calls result
```

Comment:
Why must it be memoized? The engine can handle registering the same rule multiple times.

Reply:
Only if the rules are identical, which they aren't in this case.